saleor/product/models.py

```diff
@@ -315,6 +319,10 @@ def save(
         # Make sure the "minimal_variant_price_amount" is set
         if self.minimal_variant_price_amount is None:
             self.minimal_variant_price_amount = self.price_amount
+
+        # Product can't be published if doesn't have category
+        if not self.category and self.is_published:
```

Comment:
If we have the right checks in mutations and views that modify the category model, this logic here won't be necessary. We usually don't put much constraints in the `save` method in models.